examples/repl.rs

```diff
@@ -431,6 +425,143 @@ mod example {
+                regs.rsp as usize,
+                regs.rbp as usize,
+            )
+            .collect(),
+            BtType::Naive => headcrab::symbol::unwind::naive_unwinder(
+                context.debuginfo(),
+                &stack[..],
+                regs.rip as usize,
+            )
+            .collect(),
+        };
+        Ok(call_stack)
+    }
+
+    /// Get the call_stack from the context and then tries to display the
+    /// source for the top call in the stack. Becasue the first frame is usually
+    /// sse2.rs, we just display the file and line but not the source and we skip
+    /// over to the next frame. For the next frame, we will display the source code.
+    /// An example view is shown below:
+    ///
+    /// It marks the line with the berakpoint with a '>' character and shows some lines
+    /// of context above and below it.
+    ///
+    /// 0000555555559295 core::core_arch::x86::sse2::_mm_pause /../rustup/toolchains/1.45.2-x86_64-unknown-linux-gnu/../stdarch/crates/core_arch/src/x86/sse2.rs:25
+    /// /workspaces/headcrab/tests/testees/hello.rs:7:14
+    ///    4 #[inline(never)]
+    ///    5 fn breakpoint() {
+    ///    6     // This will be patched by the debugger to be a breakpoint
+    /// >  7     unsafe { core::arch::x86_64::_mm_pause(); }
+    ///    8 }
+    ///    9
+    ///   10 #[inline(never)]
+    fn print_source_for_top_of_stack_symbol(
+        context: &mut Context,
+        context_lines: u32,
+    ) -> Result<(), Box<dyn std::error::Error>> {
+        let call_stack = get_call_stack(context, BtType::FramePtr)?;
+        let top_of_stack = call_stack[0];
+        context
+            .debuginfo()
+            .with_addr_frames(top_of_stack, |_addr, mut frames| {
+                let mut first_frame = true;
+                while let Some(frame) = frames.next()? {
+                    let name = frame
+                        .function
+                        .as_ref()
+                        .map(|f| Ok(f.demangle()?.into_owned()))
+                        .transpose()
+                        .map_err(|err: gimli::Error| err)?
+                        .unwrap_or_else(|| "<unknown>".to_string());
+
+                    let (file, line, column) = frame
+                        .location
+                        .as_ref()
+                        .map(|loc| {
+                            (
+                                loc.file.unwrap_or("<unknown file>"),
+                                loc.line.unwrap_or(0),
+                                loc.column.unwrap_or(0),
+                            )
+                        })
+                        .unwrap_or_default();
+
+                    if first_frame {
+                        println!(
+                            "{:016x} {} {}",
+                            top_of_stack,
+                            name,
+                            format!("{}:{}", file, line)
+                        );
```

Comment:
This is wrong when you stepped out of `_mm_pause`. Maybe just specifically skip any function with `_mm_pause` in the name. (And add a FIXME to remove it once `_patch_breakpoint_function` is gone)

Reply:
makes sense. Done !